File: backend/app/schemas/trace.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class AgentTraceResponse(BaseModel):
    id: int
    session_id: str | None
    thread_id: str | None
    problem_slug: str | None
    node_name: str
    phase: str | None
    hint_level: int | None
    model_name: str | None
    latency_ms: int | None
    stuck_point: str | None
    should_reveal_solution: bool | None
    input_summary: dict[str, Any] = Field(default_factory=dict)
    output_summary: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime

File: frontend/src/api/trace.ts
import { requestJson } from './client'

export type AgentTrace = {
  id: number
  session_id: string | null
  thread_id: string | null
  problem_slug: string | null
  node_name: string
  phase: string | null
  hint_level: number | null
  model_name: string | null
  latency_ms: number | null
  stuck_point: string | null
  should_reveal_solution: boolean | null
  retrieved_chunk_ids: unknown[]
  input_summary: Record<string, unknown>
  output_summary: Record<string, unknown>
  created_at: string
}

export function getAgentTraces(sessionId?: number | null) {
  const query = sessionId ? `?session_id=${sessionId}` : ''
  return requestJson<AgentTrace[]>(`/api/traces${query}`)
}
